File: python/hot100/66.搜索旋转排序数组.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1
        while left < right:
            mid = (left + right) // 2
            x = nums[mid]
            if target > nums[-1] >= x:
                right = mid
            elif x > nums[-1] >= target:
                left = mid + 1
            elif x >= target:
                right = mid
            else:
                left = mid + 1
        return right if nums[right] == target else -1

File: cpp/hot100/66.搜索旋转排序数组.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0, right = nums.size() - 1;
        int last = nums.back();
        while(left < right){
            int mid = left + (right - left) / 2;
            int x = nums[mid];
            if(target > last && x <= last){
                right = mid;
            }
            else if(target <= last && x > last){
                left = mid + 1;
            }
            else if(x >= target){
                right = mid;
            }
            else{
                left = mid + 1;
            }
        }
        return nums[right] == target ? right : -1;
    }
};